File: scripts/drive_scanner.py
"""
drive_scanner.py — Scan a Google Drive folder for creative assets.

Lists all files in a shared Drive folder and returns metadata with
direct-access URLs for images and videos.

Usage:
    from scripts.drive_scanner import build_file_id_map
    file_map = build_file_id_map("1DEF...uvw")
    # Returns: { "filename.jpg": { "file_id": "...", "image_url": "...", ... } }
"""
from scripts.google_auth import get_drive_service


# File extensions we care about
IMAGE_EXTS = {"jpg", "jpeg", "png", "gif", "webp"}
VIDEO_EXTS = {"mp4", "mov"}
ALL_CREATIVE_EXTS = IMAGE_EXTS | VIDEO_EXTS

# MIME type to extension mapping
MIME_TO_EXT = {
    "image/jpeg": "jpg",
    "image/png": "png",
    "image/gif": "gif",
    "image/webp": "webp",
    "video/mp4": "mp4",
    "video/quicktime": "mov",
}


def _make_image_url(file_id):
    """Direct view URL for images (renders in browser/img tag)."""
    return f"https://drive.google.com/uc?export=view&id={file_id}"


def _make_download_url(file_id):
    """Direct download URL for videos."""
    return f"https://drive.google.com/uc?export=download&id={file_id}"


def scan_drive_folder(folder_id, service=None):
    """
    List all creative files in a Google Drive folder.

    Returns a list of dicts:
    [
        {
            "filename": "Edgemont_BrookridgeLane_Meta_Single_Feb2026.jpg",
            "file_id": "1a2b3c4d...",
            "mime_type": "image/jpeg",
            "ext": "jpg",
            "is_video": False,
            "image_url": "https://drive.google.com/uc?export=view&id=...",
            "download_url": "https://drive.google.com/uc?export=download&id=...",
        },
        ...
    ]
    """
    if service is None:
        service = get_drive_service()

    files = []
    page_token = None

    while True:
        response = (
            service.files()
            .list(
                q=f"'{folder_id}' in parents and trashed = false",
                fields="nextPageToken, files(id, name, mimeType, size)",
                pageSize=100,
                pageToken=page_token,
                orderBy="name",
            )
            .execute()
        )

        for f in response.get("files", []):
            name = f["name"]
            ext = name.rsplit(".", 1)[-1].lower() if "." in name else ""

            # Skip non-creative files
            if ext not in ALL_CREATIVE_EXTS:
                continue

            file_id = f["id"]
            is_video = ext in VIDEO_EXTS

            files.append(
                {
                    "filename": name,
                    "file_id": file_id,
                    "mime_type": f.get("mimeType", ""),
                    "ext": ext,
                    "is_video": is_video,
                    "image_url": _make_image_url(file_id),
                    "download_url": _make_download_url(file_id),
                    "size": int(f.get("size", 0)),
                }
            )

        page_token = response.get("nextPageToken")
        if not page_token:
            break

    return files


def build_file_id_map(folder_id, service=None):
    """
    Scan a Drive folder and return a filename → metadata dict.

    Returns:
    {
        "Edgemont_BrookridgeLane_Meta_Single_Feb2026.jpg": {
            "file_id": "1a2b3c4d...",
            "image_url": "https://drive.google.com/uc?export=view&id=...",
            "download_url": "https://drive.google.com/uc?export=download&id=...",
            "is_video": False,
            "ext": "jpg",
        },
        ...
    }
    """
    files = scan_drive_folder(folder_id, service)
    return {
        f["filename"]: {
            "file_id": f["file_id"],
            "image_url": f["image_url"],
            "download_url": f["download_url"],
            "is_video": f["is_video"],
            "ext": f["ext"],
        }
        for f in files
    }


def print_drive_summary(folder_id, service=None):
    """Print a summary of files found in a Drive folder."""
    files = scan_drive_folder(folder_id, service)
    images = [f for f in files if not f["is_video"]]
    videos = [f for f in files if f["is_video"]]

    print(f"Drive folder: {folder_id}")
    print(f"Total creative files: {len(files)}")
    print(f"  Images: {len(images)}")
    print(f"  Videos: {len(videos)}")
    print()

    for f in files:
        size_kb = f["size"] / 1024 if f["size"] else 0
        kind = "VIDEO" if f["is_video"] else "IMAGE"
        print(f"  [{kind}] {f['filename']} ({size_kb:.0f} KB)")

File: src/utils/driveScanner.js
/**
 * driveScanner.js — Scan a Google Drive folder for creative assets.
 *
 * Browser port of scripts/drive_scanner.py.
 * Uses fetch() with OAuth2 access token instead of service account.
 *
 * Recursively scans subfolders so users can point at a parent folder
 * that contains per-campaign or per-placement sub-folders.
 */
import { googleFetch } from './googleAuth';

const IMAGE_EXTS = new Set(['jpg', 'jpeg', 'png', 'gif', 'webp']);
const VIDEO_EXTS = new Set(['mp4', 'mov']);
const ALL_CREATIVE_EXTS = new Set([...IMAGE_EXTS, ...VIDEO_EXTS]);
const FOLDER_MIME = 'application/vnd.google-apps.folder';

function makeImageUrl(fileId) {
  return `https://drive.google.com/uc?export=view&id=${fileId}`;
}

function makeDownloadUrl(fileId) {
  return `https://drive.google.com/uc?export=download&id=${fileId}`;
}

/**
 * List all items (files + folders) in a single Drive folder.
 */
async function listFolder(folderId) {
  const items = [];
  let pageToken = null;

  do {
    const q = encodeURIComponent(`'${folderId}' in parents and trashed = false`);
    const fields = encodeURIComponent('nextPageToken, files(id, name, mimeType, size)');
    let url = `https://www.googleapis.com/drive/v3/files?q=${q}&fields=${fields}&pageSize=100&orderBy=name`;
    if (pageToken) {
      url += `&pageToken=${encodeURIComponent(pageToken)}`;
    }

    const response = await googleFetch(url);
    items.push(...(response.files || []));
    pageToken = response.nextPageToken || null;
  } while (pageToken);

  return items;
}

/**
 * Recursively scan a Google Drive folder for creative files.
 *
 * Returns a filename → metadata map (same as Python's build_file_id_map).
 *
 * @param {string} folderId - Google Drive folder ID
 * @param {function} onProgress - Progress callback
 */
export async function buildFileIdMap(folderId, onProgress = null) {
  const fileMap = {};
  let scannedFolders = 0;

  async function scanFolder(id, depth = 0) {
    scannedFolders++;
    onProgress?.(`Scanning Drive folder${scannedFolders > 1 ? `s (${scannedFolders} scanned)` : ''}…`);

    const items = await listFolder(id);
    console.log(`[DriveScanner] Folder ${id} (depth ${depth}): ${items.length} items`, items.map(f => `${f.name} (${f.mimeType})`));

    for (const f of items) {
      // Recurse into subfolders (max depth 5 to avoid runaway)
      if (f.mimeType === FOLDER_MIME) {
        if (depth < 5) {
          await scanFolder(f.id, depth + 1);
        }
        continue;
      }

      const name = f.name;
      const ext = name.includes('.') ? name.split('.').pop().toLowerCase() : '';

      // Skip non-creative files
      if (!ALL_CREATIVE_EXTS.has(ext)) continue;

      const fileId = f.id;
      const isVideo = VIDEO_EXTS.has(ext);

      fileMap[name] = {
        file_id: fileId,
        image_url: makeImageUrl(fileId),
        download_url: makeDownloadUrl(fileId),
        is_video: isVideo,
        ext,
      };
    }
  }

  await scanFolder(folderId);
  const fileCount = Object.keys(fileMap).length;
  console.log(`[DriveScanner] DONE: ${fileCount} creative files in ${scannedFolders} folders`, Object.keys(fileMap));
  onProgress?.(`Found ${fileCount} creative files in ${scannedFolders} folder${scannedFolders > 1 ? 's' : ''}.`);

  return fileMap;
}
